calibre-plugin/keyextract/main.c

```diff
@@ -1,196 +1,112 @@
-    }
-    if (len < 0) {
-        fprintf(stderr, "  NEGATIVE LENGTH: %d\n", len);
-        return;
-    }
-
-    // Process every byte in the data.
-
-    for (i = 0; i < len; i++) {
-        // Multiple of perLine means new or first line (with line offset).
-
-        if ((i % perLine) == 0) {
-            // Only print previous-line ASCII buffer for lines beyond first.
-
-            if (i != 0) fprintf (stderr, "  %s\n", buff);
-
-            // Output the offset of current line.
+	DWORD be_serial = htonl(serial);
+	fprintf(stderr, "Disk serial (hex): %08lx\n", serial);
 
-            fprintf (stderr, "  %04x ", i);
-        }
 
-        // Now the hex code for the specific character.
-
-        fprintf (stderr, " %02x", pc[i]);
-
-        // And buffer a printable ASCII character for later.
-
-        if ((pc[i] < 0x20) || (pc[i] > 0x7e)) // isprint() may be better.
-            buff[i % perLine] = '.';
-        else
-            buff[i % perLine] = pc[i];
-        buff[(i % perLine) + 1] = '\0';
-    }
-
-    // Pad out last line if not exactly perLine characters.
-
-    while ((i % perLine) != 0) {
-        fprintf (stderr, "   ");
-        i++;
-    }
-
-    // And print the final ASCII buffer.
-
-    fprintf (stderr, "  %s\n", buff);
-}
-#endif
-
-int get_serial() {
-	DWORD serial = 0;
-	int retval = GetVolumeInformation("c:\\\\", NULL, 0, &serial, NULL, NULL, NULL, 0);
-	if (retval == 0) {
-		fprintf(stderr, "Error with GetVolumeInformation: %d\n", GetLastError());
-		return 0;
+	unsigned int eax, ebx, ecx, edx;
+	// Get CPUID vendor string
+	union CPUIDVendor cpu_vendor;
+	if (__get_cpuid(0, &eax, &cpu_vendor.reg[0], &cpu_vendor.reg[2], &cpu_vendor.reg[1]) == 0) {
+		fprintf(stderr, "Error: cpuid(0) not supported");
+		return 1;
 	}
-	return serial;
-}
-
-int main() {
-	char * var_data = "X_DECRYPT_DATA";
-	char * var_entropy = "X_DECRYPT_ENTROPY";
-
-	char * data_hex = getenv(var_data);
-	char * entropy_hex = getenv(var_entropy);
-
-	if (data_hex == NULL || entropy_hex == NULL) {
-		fputs("PROGOUTPUT:-1", stdout);
-		exit(-1);
+	cpu_vendor.vendor[12] = '\0';
+	fprintf(stderr, "CPUID Vendor: %s\n", cpu_vendor.vendor);
+	// Get CPUID "signature" (eax of CPUID(1))
+	unsigned int signature;
+	if (__get_cpuid(1, &signature, &ebx, &ecx, &edx) == 0) {
+		fprintf(stderr, "Error: cpuid(1) not supported");
+		return 1;
 	}
-
-	char * data_bytes = malloc((strlen(data_hex) / 2));
-	char * entropy_bytes = malloc((strlen(entropy_hex) / 2));
-
-	if (data_bytes == NULL || entropy_bytes == NULL) {
-		fputs("PROGOUTPUT:-2", stdout);
-		exit(-2);
+	unsigned int be_signature = htonl(signature);
+	fprintf(stderr, "CPUID Signature (hex): %08x\n", signature);
+
+
+	// Get windows user
+	#define USERBUFSIZE 512
+	TCHAR user[USERBUFSIZE];
+	memset(&user, 0, sizeof(user));  // GetUserName only sets bytes as needed for length of username, but we need null bytes to fill the rest
+	DWORD bufsize = USERBUFSIZE	;
+	if (GetUserName(user, &bufsize) == 0) {
```

Comment:
I see - the wine registry might have been created with a different username. In that case, I would assume that `GetUserName` would return the username that the wine context is user, but I'm not sure of that since wine [doesn't document that function](https://source.winehq.org/WineAPI/G.html). I'll switch it to get it from `Volatile Environment\USERNAME` instead as was done before.

Reply:
No, don't do that. The issue is independantly from Wine and would happen on real Windows, too. 
You'll need to read the username from `Software\\\\Adobe\\\\Adept\\\\Device` if available, and only fall back to GetUserName if that doesn't exist. 

ADE itself decides to use the username (at time of authorization) as part of the encryption key, and it then stores the username in the registry at the key mentioned above, in case the username gets changed afterwards. ADE will always use the username it stored into the registry itself, instead of the one whatever Windows API returns.

Reply:
I chose `Volatile Environment\USERNAME` because that was what was being used before, along with `Software\\Adobe\\Adept\\Device`, though I think I understand the difference now. I'll keep using `GetUserName` as a fallback in that case and read the adobe key for the username.

Reply:
"Volatile Environment" was only used if the Adobe key didn't exist.

Reply:
Ok, reading the `username` key works and so does the fallback if it doesn't exist. Is it possible for the key to be missing at all in the first place, apart from manual removal?

Reply:
I am not sure. I've had that happen during my testing (so I added the fallback), but I'm not sure if it was actually missing or if I just had fucked it up while messing with the registry, so I thought it couldn't hurt to leave the fallback in, even though it might not be necessary.